### ISSUE TITLE
Add setting to `Style/CaseEquality` to allow `x.===(y)` (explicit parentheses)

### ISSUE THREAD

**Issue by jez** [rubocop/rubocop#14684]
## Is your feature request related to a problem? Please describe.

```ruby
/some_pattern/ === x
```

This code is rejected by the `Style/CaseEquality` cop. The recommended style here requires doing

```ruby
case x when /some_pattern/ then true else false end
```

which is more verbose.

One of the main motivations for banning `===` is to ensure that people who frequently work with JavaScript doesn't use `===` out of habit, mistakenly assuming that it is the "normal" equality operator of `==`.

However, there are valid use cases for `===`. For example, if you tried to rewrite `/some_pattern/ === x` without, you either end up with the verbose solution above, or the more brittle solution of

```ruby
x =~ /some_pattern/
```

which will raise an exception if `x` doesn't have the `=~` method, or

```ruby
/some_pattern/ =~ x
```

which will raise an exception if `x` cannot be implicitly converted into a `String`.

In this case, `/some_pattern/ === x` is the least brittle and most concise solution to the problem. 

## Describe the solution you'd like

It would be nice to have an option for the `Style/CaseEquality` cop that permits the call if it uses the explicit `x.===(y)` syntax, i.e., an explicit `.` and parentheses.

- It's unlikely that people switching between a JavaScript and Ruby codebase would habitually write this form—it would be far more likely that they intend to call Ruby's case equality operator, not generic equality.
- It's still more terse than the alternative of fully expanding a `case/when` expression.

This would be guarded with an option (like the two existing options for the `Style/CaseEquality` cop: `AllowOnConstant` and `AllowOnSelfClass`).

As a suggestion: `AllowOnExplicitCall`, but I'm open to alternative names.

## Describe alternatives you've considered

There are some alternatives considered above, like

- just use `case/when`
- just disable the cop when this is needed
- codebases could define a separate helper method, like `Utils.case_equality(x, y)` to make things more explicit

## Additional context

The concrete problem we're encountering is when upgrading legacy Ruby codebases that relied on the existence of `Object#=~`, which was made deprecated and eventually removed. As we find instances where the code was relying on this, we would like to be able to use `===` to replace the call to `=~` but we run against this cop.

- - - - -

If adding such an option is a welcome change, I'd be happy to contribute this change.

**Comment by koic** [comment]
As described in the existing documentation, `/regexp/ === var` can potentially be replaced with `/regexp/.match?(var)`. However, these two are not compatible when captures such as `$1` are used. The `Performance/Regexp` cop exists to handle such cases.

I opened #14686 to update the documentation so that this cop ignores potentially incompatible uses of regexp `===` and delegates the handling to the `Performance/Regexp` cop.

**Comment by jez** [comment]
Thanks, though I still think I would prefer an option to use explicit parens. Another example is sometimes mapping a block over a collection:

    xs.filter { A === it }

When xs is an array, this can be changed to

    xs.grep(A)

but sometimes `xs` is some user-defined object which means that grep is not available. In such cases you have to write out the whole case statement still

    xs.filter { case it when A then true else false }

And in this case, I would want to be able to use explicit parentheses to be allowed to call === directly.

**Comment by koic** [comment]
Since `x.===(y)` goes against the style recommended in the following guideline, adding it as a RuboCop option seems uncertain:
https://rubystyle.guide/#operator-method-call

Instead of writing `xs.filter { case it when A then true else false }`, it seems more appropriate to replace it with a block that uses a suitable method rather than `case` expression, such as `xs.filter { it.is_a?(A) }`, `xs.filter { it.include?(A) }`, or `xs.filter { it.match?(A) }`.

**Comment by jez** [comment]
Thank you for the suggestion but none of those methods exist on BasicObject. Using `A.===(it)` works even if `it` does not respond to any of those methods, or if those methods have been monkey patched or stubbed to do weird things for certain classes.

I understand that you only want to build rubocop options that conform to that style guide. In that case, I will probably build an alternative rule so that I can disable the default Style/CaseEquality rule and enable an alternative CaseEquality rule in the codebases I work in.

Thanks for the responses.
